### ISSUE TITLE
Passing CMake arguments to package components

### ISSUE THREAD

**Issue by tzemaitis** [ruslo/hunter#327]
Hello,

I want to build static libraries with position independent code enabled (I need this because these static libraries at final step are compiled to single shared object). I have to pass -fPIC compiler flag to every library.

To achieve this I am passing CMake arguments with `hunter_config` function, like this:
```
hunter_config(OpenCV VERSION 3.0.0-p6 CMAKE_ARGS CMAKE_VERBOSE_MAKEFILE=ON CMAKE_POSITION_INDEPENDENT_CODE=ON)
```
But this do not work in case if package has components. I do like this:
```
hunter_config(Boost VERSION 1.58.0-p1 CMAKE_ARGS CMAKE_VERBOSE_MAKEFILE=ON CMAKE_POSITION_INDEPENDENT_CODE=ON)
```
But these arguments are not forwarded to components line Boost::regex.

Is it possible? If yes, how I can do it?

Thanks.

**Comment by ruslo** [comment]
TL;DR use toolchain to set `-fPIC` explicitly to `CMAKE_CXX_FLAGS`, e.g.:
* gcc-pic: https://github.com/ruslo/polly/blob/master/gcc-pic.cmake
* used flag: https://github.com/ruslo/polly/blob/master/flags/fpic.cmake

> But these arguments are not forwarded to components line Boost::regex

This is because Boost use custom non-CMake build scheme. It means you have to detect `CMAKE_POSITION_INDEPENDENT_CODE` yourself and add `-fPIC` to compiler flags (if it's supported, e.g. for MSVC it doesn't make sense). See related feature-request:
* [Apply CMAKE_POSITION_INDEPENDENT_CODE-like variables](https://github.com/ruslo/hunter/issues/322)

>  I have to pass -fPIC compiler flag to every library

See this feature-request:
* [HUNTER_POSITION_INDEPENDENT_CODE](https://github.com/ruslo/hunter/issues/64)

And just related topic:
* [Shared lib projects can't use Hunter on Linux](https://github.com/ruslo/hunter/issues/63)

**Comment by tzemaitis** [comment]
Hello,
Tried to use gcc-pic: https://github.com/ruslo/polly/blob/master/gcc-pic.cmake toolchain to compile boost regex library and link it so shared object. But looks like position independent code is still not applied to regex static library.

Build regex static lib logs:
```
loading initial cache file /home/jk-linuxslave/.hunter/_Base/02dd55f/0c6df42/87989ab/Build/Boost/__regex/args.cmake
-- [polly] Used toolchain: gcc / PIC / c++11 support
-- The C compiler identification is GNU 4.8.4
-- The CXX compiler identification is GNU 4.8.4
-- Check for working C compiler: /usr/bin/gcc
...
common.mkdir bin.v2/libs/regex/build/gcc-4.8/release/link-static/threading-multi
gcc.compile.c++ bin.v2/libs/regex/build/gcc-4.8/release/link-static/threading-multi/c_regex_traits.o
gcc.compile.c++ bin.v2/libs/regex/build/gcc-4.8/release/link-static/threading-multi/cpp_regex_traits.o
gcc.compile.c++ bin.v2/libs/regex/build/gcc-4.8/release/link-static/threading-multi/fileiter.o
gcc.compile.c++ bin.v2/libs/regex/build/gcc-4.8/release/link-static/threading-multi/cregex.o
gcc.compile.c++ bin.v2/libs/regex/build/gcc-4.8/release/link-static/threading-multi/icu.o
gcc.compile.c++ bin.v2/libs/regex/build/gcc-4.8/release/link-static/threading-multi/instances.o
...
gcc.compile.c++ bin.v2/libs/regex/build/gcc-4.8/release/link-static/threading-multi/winstances.o
gcc.archive bin.v2/libs/regex/build/gcc-4.8/release/link-static/threading-multi/libboost_regex-mt.a
common.copy stage/lib/libboost_regex-mt.a
``` 
I do not see compiler flags, but I think is it because CMAKE_VERBOSE is not used by boost build system.

Build shared object logs:
```
linuxslave/.hunter/_Base/02dd55f/0c6df42/87989ab/Install/lib/libboost_regex-mt.a 
/usr/bin/ld: /home/jk-linuxslave/.hunter/_Base/02dd55f/0c6df42/87989ab/Install/lib/libboost_regex-mt.a(instances.o): relocation R_X86_64_32S against `.rodata' can not be used when making a shared object; recompile with -fPIC
/home/jk-linuxslave/.hunter/_Base/02dd55f/0c6df42/87989ab/Install/lib/libboost_regex-mt.a: error adding symbols: Bad value
collect2: error: ld returned 1 exit status
```
Tried to check libboost_regex-mt.a content:
```
ar -x libboost_regex-mt.a  
readelf --relocs instances.o | egrep '(GOT|PLT|JU?MP_SLOT)'
```
No output, so it is not compiled with -fPIC

Looks like it is not enough just to use this toolchain, what I am missing?

**Comment by tzemaitis** [comment]
False alarm, I found problem in my polly copy.
Thanks

**Comment by ruslo** [comment]
> I think is it because CMAKE_VERBOSE is not used by boost build system

Yes, same reason - it's custom scheme, you can use [HUNTER_STATUS_DEBUG][1].

> I found problem in my polly copy

Cool, I'll close it then. Feel free to open new issue if you have more questions.

[1]: https://github.com/ruslo/hunter/wiki/usr.variables#hunter_status_debug
